### PR TITLE
Speedups by avoiding inner classes

### PR DESCRIPTION
**Pull request by poke1024**
Turns out that, at least in CPython, declaring `class`es inside functions is quite slow.

Running `x = Range[50000]; First[Timing[Position[x, 10000]]]` without this PR gives `1.16326`, with this PR, it gives `0.461211`. This is only due to one inner `class` definition.

Running the profiler, one actually sees Python's class build function popping up. So this PR moves inner classes out of functions at several places.

### CODE REVIEW

File: mathics/builtin/patterns.py [hunk 1]
```diff
@@ -1332,3 +1334,25 @@ def _match(leaf):
...
+    try:
+        form.match(yield_match, expr, {}, evaluation, fully=False)
+    except _StopGeneratorBaseExpressionIsFree as exc:
+        return exc.value
+
+    if expr.is_atom():
+        return True
+    else:
+        return expr.head.is_free(form, evaluation) and all(
+            leaf.is_free(form, evaluation) for leaf in expr.leaves)
```
Comment (sn6uv) [inline mathics/builtin/patterns.py]: is this function related to the change?

Comment (poke1024) [inline mathics/builtin/patterns.py]: I forgot to hook this up. It's done now. I needed to move this function into `patterns.py` as the declaration of `_StopGeneratorBaseExpressionIsFree` needs `StopGenerator` which cannot be imported into `core.expression.py` as `patterns.py` depends on `core.expression.py`.
